ledger/participant-state/kvutils/tools/src/test/scala/com/daml/Replay.scala

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.test
+
+import java.io.DataInputStream
+import java.lang.System.err.println
+import java.nio.file.{Files, Path, Paths}
+import java.util.concurrent.TimeUnit
+
+import com.daml.ledger.participant.state.kvutils.Conversions._
+import com.daml.ledger.participant.state.kvutils.export.FileBasedLedgerDataExporter.SubmissionInfo
+import com.daml.ledger.participant.state.kvutils.export.Serialization
+import com.daml.ledger.participant.state.kvutils.{Envelope, DamlKvutils => Proto}
+import com.daml.ledger.participant.state.v1.ParticipantId
+import com.daml.lf.archive.{Decode, UniversalArchiveReader}
+import com.daml.lf.data._
+import com.daml.lf.engine.Engine
+import com.daml.lf.transaction.Node.GlobalKey
+import com.daml.lf.transaction.{Node, Transaction => Tx, TransactionCoder => TxCoder}
+import com.daml.lf.value.Value.ContractId
+import com.daml.lf.value.{Value, ValueCoder => ValCoder}
+import com.daml.lf.{crypto, data}
+import com.google.protobuf.ByteString
+import org.openjdk.jmh.annotations._
+
+import scala.collection.JavaConverters._
+
+case class TxEntry(
+    tx: Tx.SubmittedTransaction,
+    participantId: ParticipantId,
+    ledgerTime: Time.Timestamp,
+    submissionTime: Time.Timestamp,
+    submissionSeed: crypto.Hash,
+)
+
+case class BenchMarkSate(
+    name: String,
+    transaction: TxEntry,
+    contracts: ContractId => Option[Tx.ContractInst[ContractId]],
+    contractKeys: GlobalKey => Option[ContractId],
+)
+
+@State(Scope.Benchmark)
+class Replay {
+
+  @Param(Array())
```

Comment:
What does it mean to use an empty array as a `@Param`? Does this run anything at all?

I would have thought that the cross product of `Array.empty` × _anything_ would be `Array.empty`.

Reply:
This instruments the code such that you can use the `JMH` command line with 
`-p ledgerFile=ledger.export` (See the "usage" in the description). 
Using an empty Array, is just a wait to specify there is no default value, (I do not see any reasonable default value). In other word it obliges you to use the command line parameter "-p ".

Reply:
That's great, thanks for the explanation.